File: Codes/Programmers/다트게임.py
def solution(dartResult):
    def S():
        return

    def D():
        score[-1] **= 2

    def T():
        score[-1] **= 3

    def star():
        score[-1] *= 2
        if len(score) > 1:
            score[-2] *= 2

    def acha():
        score[-1] *= -1

    bonus_option = {"S": S, "D": D, "T": T, "*": star, "#": acha}
    score = []
    number = ""

    for char in dartResult:
        if char.isdigit():
            number += char
        else:
            if number:
                score.append(int(number))
                number = ""
            bonus_option[char]()

    return sum(score)

File: Codes/Programmers/다트게임.js
function solution(dartResult) {
  const isNumber = (num) => (48 <= num.charCodeAt()) & (num.charCodeAt() <= 57);

  const f = {
    S: () => {},
    D: () => {
      score[score.length - 1] **= 2;
    },
    T: () => {
      score[score.length - 1] **= 3;
    },
    "*": () => {
      score[score.length - 1] *= 2;
      if (score.length > 1) {
        score[score.length - 2] *= 2;
      }
    },
    "#": () => {
      score[score.length - 1] *= -1;
    },
  };

  const score = [];
  let number = "";
  for (char of dartResult) {
    if (isNumber(char)) {
      number += char;
      continue;
    }
    if (number.length) {
      score.push(+number);
      number = "";
    }
    f[char]();
  }
  return score.reduce((a, b) => a + b);
}
